utils/filename_test.go

```diff
@@ -0,0 +1,139 @@
+package utils
+
+import (
+	"net/url"
+	"testing"
+)
+
+type filenameScenario struct {
+	scenario string
+	given    string
+	expected string
+}
+
+var scenarios = []filenameScenario{
+	{
+		scenario: "empty",
+		given:    "",
+		expected: "",
+	},
+	{
+		scenario: "no escapes",
+		given:    "this_is_valid",
+		expected: "this_is_valid",
+	},
+	{
+		scenario: "spaces",
+		given:    "this has spaces",
+		expected: "this%20has%20spaces",
+	},
+	{
+		scenario: "capitals",
+		given:    "This Has Caps And Spaces",
+		expected: "%54his%20%48as%20%43aps%20%41nd%20%53paces",
+	},
+	{
+		scenario: "emojis",
+		given:    "happy😀file",
```

Comment:
emojis in filenames work just fine on windows, macos and linux, no need to escape them

Reply:
I was using emojis as an easy stand in for the overall extended unicode character set. My concern was mainly about case sensitivity in russian or greek character sets, for example.

This was a demonstration that multi-byte and multi-code point graphemes will encode and decode correctly. I'm not sure how I would do that otherwise, unless you want to bring in dedicated unicode libraries.

Reply:
I'm wondering if unicode filenames are going to be a problem or not at all in the first place.

I'm wondering if we're being excessively defensive here. Is there an example of files on repos on github that cannot be correctly `git clone`d on windows because of unicode characters that are file on linux or mac but not on windows?

Ideally we should mangle names the least amount of possible, in order to keep the vendored directory to feel as much as possible "native" and "reviewable" and reserve the escaping only for edge cases where we know people will be very confused by the behavior of the system due to illegal filenames.

I'm also trying to figure out if there is a better way to handle uppercase chars, which are going to the single most common class of escaped chars under the current scheme.

Reply:
for example, what if instead of escaping uppercases, we implement some sort of conflict resolution pass, where we sanitize the entire list of files and we only escape things when there are actual conflicts? (not sure how to do that)

Reply:
I have an update that mangles only uppercase letters, and reserved words and symbols. So lowercase greek, cyrillic, and other character sets are left alone. In general, so are symbols that are not letter-like.

Let me know if you see any problems